---
title: "10X QC processing post cellranger"
author: "Kimberly Olney, PhD"
date: "03/11/2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
# Lewy body dementia center without walls LBD CWOW

# Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = "/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/")
setwd("/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/")
```

# Libraris, paths, colors
```{r echo=FALSE, message=FALSE}
source(here::here("/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/", "file_paths_and_colours.R"))
project_ID <- "CWOW_cellbender"
```

# Genes 
```{r gene_info}
if (file.exists("../../rObjects/annotation.rds")) {
  genes <- readRDS("../../rObjects/annotation.rds")
} else {
  gtf.file <- paste0(pathToRef, "/genes/genes.gtf")
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../rObjects/annotation.rds")
}

# gene table is an overview of the types of genes in the annotation file 
gene_type_table <- table(genes$gene_type)
write.table(gene_type_table, "gene_type_table.tsv", row.names = F, quote = F, sep = "\t")
write.table(genes, "genes.tsv", row.names = F, quote = F, sep = "\t")

# obtain the MT genes. Will be used for filtering down stream. 
mt.genes.df <- subset(genes, seqnames == "chrM")
genes_mt <- mt.genes.df$gene_name # 13 Mito genes 

rm(mt.genes.df, gene_type_table)
```

# QC thresholds
nCount_RNA = total number of transcripts (UMIs) in a single cell 
nFeature_RNA = number of unique genes (features)
```{r QC_metrics}
nCount.min <- 1000 
nCount.max <- 100000 
nFeature.min <- 800
nFeature.max<- 12000
cutoff_complexity <- 0.80
cutoff_mt <- 5
cutoff_hb <- 1
cutoff_ribo <- 1
cutoff_choroid <- 1
```

# Setup seurat object
```{r seurat_object}
dataObject <- readRDS(paste0("../../rObjects/", project_ID, ".rds"))
```

# Change indents
```{r change_idents}
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
table(sample)
unique(sample)
dataObject$sample <- factor(sample)

# samples to be removed
rm_samples <- c("BR_Nuclei_0404", "BR_Nuclei_0376", "BR_Nuclei_0407", "BR_Nuclei_0394", "BR_Nuclei_0403") # BR_Nuclei_0404
# cell IDs
cells_to_remove <- rownames(dataObject@meta.data)[dataObject@meta.data$sample %in% rm_samples]
cells_to_remove_found <- intersect(cells_to_remove, colnames(dataObject))
# Check 
print(paste("Found", length(cells_to_remove_found), "cells to remove out of", length(colnames(dataObject)), "requested."))
# Use the 'cells_to_remove_found' variable to safely subset the data
dataObject_filtered <- subset(x = dataObject, cells = colnames(dataObject)[!colnames(dataObject) %in% cells_to_remove_found])

# 1. Access the metadata and apply droplevels() to the 'sample' column
dataObject_filtered$sample <- droplevels(dataObject_filtered$sample)
# 2. Verify that the unused levels are gone
table(dataObject_filtered$sample)
```

```{r}
rm(rm_samples, cells_to_remove, cells_to_remove_found, barcodes, dataObject)
dataObject <- dataObject_filtered
rm(dataObject_filtered)
```

# Reorder samples
```{r reorder_samples}
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
dataObject$sample <- factor(sample, levels = order_samples)
table(dataObject$sample)  # check
Idents(dataObject) <- dataObject$sample
rm(sample, barcodes)
```

# Add metadata
```{r}
# Assuming 'metadata' is your metadata dataframe
# and its 'sampleID' column contains the cell IDs (the same IDs as the Seurat object's cells)
rownames(metadata) <- metadata$sampleID
# Reorder the metadata data frame to match the column order of the Seurat object
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
metadata_reordered <- metadata[sample, ]
# Add the entire metadata dataframe to the Seurat object
# Add the reordered metadata to the Seurat object
dataObject <- AddMetaData(object = dataObject, metadata = metadata_reordered)
dataObject$group <- factor(dataObject$TYPE, levels = c("CONTROL", "AD_AT", "LBD_S", "LBD_AS", "LBD_ATS"))
```

nuclei count per sample, sex, and group
```{r}
# Sample
table(dataObject$Sample_ID) 

# Sex 
table(dataObject$sex)

# Type
table(dataObject$group) 
```

QC columns 
```{r QC_columns}
summary(dataObject$nCount_RNA)
summary(dataObject$nFeature_RNA)
# cell.complexity
dataObject$cell.complexity <- log10(dataObject$nFeature_RNA) / log10(dataObject$nCount_RNA)

# Chromosome M
gene.names <- rownames(dataObject)
dataObject$percent.mt <- PercentageFeatureSet(dataObject, features = genes_mt)
summary(dataObject$percent.mt)

"^RPS|^RPL"
# ribosomal proteins 
ribo.genes <- gene.names[grep("^RPS|^RPL", gene.names)] 
mt.ribo <- gene.names[grep("^MRP[SL]", gene.names)]
ribo.combined <- c(mt.ribo,ribo.genes)
dataObject$percent.ribo <- PercentageFeatureSet(dataObject, features = ribo.combined)
summary(dataObject$percent.ribo)

# hemoglobin proteins
hb.genes <- gene.names[grep("^HB[BA]", gene.names)]
dataObject$percent.hb <- PercentageFeatureSet(dataObject, features = hb.genes)
summary(dataObject$percent.hb)

# percent choroid plexus
dataObject$percent.choroid <- PercentageFeatureSet(dataObject, features = c("TTR","FOLR1", "PRLR"))
summary(dataObject$percent.choroid)
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject$Sample_ID))
colnames(data) <- c("Sample_ID","frequency")

ncells1 <- ggplot(data, aes(x = Sample_ID, y = frequency, fill = Sample_ID)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            #vjust=-0.25, 
            hjust = -.025,
            angle = 90) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,30000, by = 2000), limits = c(0,30000)) +
  ggtitle("Raw: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1

rm(data)
```

## Density plots
```{r prefiltering_density}
# set graphical parameter
par(mfrow = c(7,1))
# Visualize nCount_RNA
den1 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nCount_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) + theme(legend.position="none") +
  annotate("text", x = nCount.min, y = 0.1, label = paste("Min:", nCount.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nCount.max, y = 0.1, label = paste("Max:", nCount.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize nFeature
den2 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nFeature_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min) +
  geom_vline(xintercept = nFeature.max) + theme(legend.position="none") +
  annotate("text", x = nFeature.min, y = 0.1, label = paste("Min:", nFeature.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nFeature.max, y = 0.1, label = paste("Max:", nFeature.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = cell.complexity,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = cutoff_complexity) + theme(legend.position="none")  +
  annotate("text", x = cutoff_complexity, y = 0.1, label = paste("Min:", cutoff_complexity), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent.mt
den4 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.mt,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_mt) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_mt, y = 0.1, label = paste("Max:", cutoff_mt), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.ribo
den5 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.ribo,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_ribo) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_ribo, y = 0.1, label = paste("Max:", cutoff_ribo), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.hb
den6 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.hb,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_hb) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_hb, y = 0.1, label = paste("Max:", cutoff_hb), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


den7 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.choroid,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_choroid) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Choroid Marker Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_choroid, y = 0.1, label = paste("Max:", cutoff_choroid), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Arrange graphs in grid
plots1 <- list(den1,den2,den3,den4,den5,den6,den7)
layout1 <- rbind(c(1),c(2),c(3),c(4),c(5),c(6),c(7))
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

```{r}
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
path <- paste0("../../results/density/",project_ID,"_density_raw")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 10)
dev.off()

rm(grid1, plots1)
```

## Violin plots
Warning: Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead."

This warning tells you that Seurat, by default, prefers to plot from the data (normalized) slot. However, we haven't normalized the RNA assay, thus the data slot is empty. As a fallback, Seurat  switched to using the counts slot, which contains the raw UMI counts. 
```{r prefiltering_violins}
# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(dataObject,
              features = c( "nCount_RNA","nFeature_RNA", "cell.complexity"),
              ncol = 1,
              group.by = 'Sample_ID',
              ##cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(dataObject,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 2,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v2
```

```{r, echo=FALSE}
# save v1
v1
path <- paste0("../../results/violin/",project_ID,"_nFeature_nCount_complexity_raw")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 9)
dev.off()

# save v2
v2
path <- paste0("../../results/violin/",project_ID,"_percent_raw")
saveToPDF(paste0(path, ".pdf"), width = 16, height = 7)
dev.off()

# cleanup
#remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter1, warning=FALSE}
s1 <- ggplot(
  dataObject@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_vline(xintercept = nCount.max) + 
  geom_hline(yintercept = nFeature.min) + 
  geom_hline(yintercept = nFeature.max) + 
  facet_wrap(~Sample_ID, ncol = 7) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1
```

```{r echo=FALSE}
# save
s1
path <- paste0("../../results/scatter/",project_ID,"_nFeature_vs_nCount_percentMt_raw")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 15)
dev.off()

# cleanup
remove(s1)
```

```{r prefiltering_scatter2}
s2 <- FeatureScatter(dataObject,
               feature1 = "nCount_RNA",
               feature2 = "percent.mt",
               group.by = 'Sample_ID',
               #cols = sample_colors,
               shuffle = TRUE)
s2
```

```{r,echo=FALSE}
# save
s2
path <- paste0("../../results/scatter/",project_ID,"_nCount_vs_percentMT_raw")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# cleanup
remove(s2)
```

# Filtering
## Cell-level filtering
We want to be careful filtering because removing things can easily lead to
misinterpretation. 
```{r}
# 1. Initial number of nuclei
initial_nuclei <- ncol(dataObject)
message(paste0("Initial number of nuclei: ", initial_nuclei))

# Create a working copy of the object
filtered_data <- dataObject

# Define your thresholds (ensure these are set based on your VlnPlot analysis)
# Already defined at the top of the script for easy referencing. 

# --- Apply Filters Sequentially and Report ---
# Filter 1: nCount_RNA
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = nCount_RNA > nCount.min & nCount_RNA < nCount.max)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on nCount_RNA (new total: ", ncol(filtered_data), ")"))

# Filter 2: nFeature_RNA
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = nFeature_RNA > nFeature.min & nFeature_RNA < nFeature.max)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on nFeature_RNA (new total: ", ncol(filtered_data), ")"))

# Filter 3: cell.complexity (assuming this metric is pre-calculated in your object)
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = cell.complexity > cutoff_complexity)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on cell.complexity (new total: ", ncol(filtered_data), ")"))

# Filter 4: percent.mt
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = percent.mt <= cutoff_mt)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on percent.mt (new total: ", ncol(filtered_data), ")"))

# Filter 5: percent.hb
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = percent.hb <= cutoff_hb)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on percent.hb (new total: ", ncol(filtered_data), ")"))

# Filter 6: percent.ribo
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = percent.ribo <= cutoff_ribo)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on percent.ribo (new total: ", ncol(filtered_data), ")"))

# Filter 7: percent.choroid
prev_nuclei <- ncol(filtered_data)
filtered_data <- subset(filtered_data, subset = percent.choroid <= cutoff_choroid)
removed_count <- prev_nuclei - ncol(filtered_data)
message(paste0("Removed ", removed_count, " nuclei based on percent.choroid (new total: ", ncol(filtered_data), ")"))


# Final filtered object
dataObject.filtered <- filtered_data
final_nuclei <- ncol(dataObject.filtered)
message(paste0("\nTotal nuclei remaining after all filters: ", final_nuclei))
message(paste0("Total nuclei removed: ", initial_nuclei - final_nuclei))


# Compare counts per Sample_ID before and after
message("\nNuclei count per Sample_ID (Original):")
print(table(dataObject$Sample_ID))
message("\nNuclei count per Sample_ID (Filtered):")
print(table(dataObject.filtered$Sample_ID))
```


```{r cell_filtering, eval = FALSE}
# filter
dataObject.filtered <- subset(dataObject,
                        subset = (nCount_RNA > nCount.min)&
                          (nCount_RNA < nCount.max) & 
                          (nFeature_RNA > nFeature.min) & 
                          (nFeature_RNA < nFeature.max) & 
                          (cell.complexity > cutoff_complexity) &
                          (percent.mt <= cutoff_mt) & 
                          (percent.hb <= cutoff_hb) &
                          (percent.ribo <= cutoff_ribo) &
                          (percent.choroid <= cutoff_choroid))

# print nuclei removed
print(paste0(dim(dataObject)[2] - dim(dataObject.filtered)[2]," nuclei removed"))

table(dataObject$Sample_ID)
table(dataObject.filtered$Sample_ID)
```

## Save filtering
```{r,echo=FALSE}
saveRDS(dataObject.filtered, paste0("../../rObjects/",project_ID,"_filtered.rds"), compress = FALSE)
#dataObject.filtered <- readRDS(paste0("../../rObjects/",project_ID,"_filtered.rds"))
```

# Post-filtering
## Number of cells
```{r number_cells2}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject.filtered$Sample_ID))
colnames(data) <- c("Sample_ID","frequency")

ncells2 <- ggplot(data, aes(x = Sample_ID, y = frequency, fill = Sample_ID)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            #vjust=-0.25, 
            hjust = -.025,
            angle = 90) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,30000, by = 2000), limits = c(0,30000)) +
  ggtitle("Filtered: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells2
# Arrange graphs in grid
plots2 <- list(ncells1,ncells2)
layout2 <- rbind(c(1),c(2))
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)

mean(data$frequency)
median(data$frequency)
```

```{r,echo=FALSE}
# save
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)
path <- paste0("../../results/nuclei_count/",project_ID, 
               "_cells_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 8)


# cleanup
remove(ncells1,ncells2,plots2,layout2,grid2)
```

## Density plots
```{r postfiltering_density}
# set graphical parameter
par(mfrow = c(7,2))

den8 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = nCount_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) + theme(legend.position="none") +
  annotate("text", x = nCount.min, y = 0.1, label = paste("Min:", nCount.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nCount.max, y = 0.1, label = paste("Max:", nCount.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize nFeature
den9 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = nFeature_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min) +
  geom_vline(xintercept = nFeature.max) + theme(legend.position="none") +
  annotate("text", x = nFeature.min, y = 0.1, label = paste("Min:", nFeature.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nFeature.max, y = 0.1, label = paste("Max:", nFeature.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize cell complexitydataObject.filtered.split.doublets
# Quality cells are usually above 0.85
den10 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = cell.complexity,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = cutoff_complexity) + theme(legend.position="none")  +
  annotate("text", x = cutoff_complexity, y = 0.1, label = paste("Min:", cutoff_complexity), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent.mt
den11 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = percent.mt,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_mt) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_mt, y = 0.1, label = paste("Max:", cutoff_mt), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.ribo
den12 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = percent.ribo,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_ribo) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_ribo, y = 0.1, label = paste("Max:", cutoff_ribo), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.hb
den13 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = percent.hb,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_hb) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_hb, y = 0.1, label = paste("Max:", cutoff_hb), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


den14 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = percent.choroid,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_choroid) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Choroid Marker Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_choroid, y = 0.1, label = paste("Max:", cutoff_choroid), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Arrange graphs in grid
plots3 <- list(den1,den2,den3,den4,den5,den6,den7,den8,den9,den10,den11,den12,den13,den14)
layout3 <- rbind(c(1,8),c(2,9),c(3,10), c(4,11),c(5,12),c(6,13),c(7,14))
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
```

```{r, echo=FALSE}
# save
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
path <- paste0("../../results/density/",project_ID, 
               "_density_filtered")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 10)

# cleanup
remove(den1,den2,den3,den4,den5,den6,den7,den8,den9,den10,den11,den12,den13,den14,plots3,layout3,grid3)
```

## Violin plots
```{r postfiltering_violins}
# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(dataObject.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 1,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(dataObject.filtered,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 2,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v4
```

```{r, echo=FALSE}
# save
v3
path <- paste0("../../results/violin/",project_ID, 
               "_nFeature_nCount_complexity_filtered")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 11)

v4
path <- paste0("../../results/violin/",project_ID, 
               "_percent_filtered")
saveToPDF(paste0(path, ".pdf"), width = 14, height = 9)

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter1}
s3 <- ggplot(
  dataObject.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_vline(xintercept = nCount.max) + 
  geom_hline(yintercept = nFeature.min) + 
  geom_hline(yintercept = nFeature.max) + 
  facet_wrap(~Sample_ID, ncol = 8) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3
```

```{r,echo=FALSE}
# save
s3
path <- paste0("../../results/scatter/",project_ID,
               "_nFeature_vs_nCount_perecentMt_filtered")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 15)

# cleanup
remove(s3)
```

```{r postfiltering_scatter2}
s4 <- FeatureScatter(dataObject.filtered,
               feature1 = "nCount_RNA",
               feature2 = "percent.mt",
               group.by = 'Sample_ID',
               #cols = sample_colors,
               shuffle = TRUE)
s4
```

```{r,echo=FALSE}
# save
s4
path <- paste0("../../results/scatter/",project_ID,
               "_nCount_vs_percentMT_filtered")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# cleanup
remove(s4)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(dataObject.filtered@meta.data,
       aes(x = Sample_ID, 
           y = log10(nFeature_RNA), 
           fill=Sample_ID)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        legend.position = "none") +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Nuclei / Sample") +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Sample")
b1
```

```{r,echo=FALSE}
# save
b1
path <- paste0("../../results/nuclei_count/",project_ID,
               "_nFeature_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)

# cleanup
remove(b1)
```

## Top transcripts
```{r top_transcripts}
df <- data.frame(row.names = rownames(dataObject.filtered))
df$rsum <- rowSums(x = dataObject.filtered, slot = "counts")
df$gene_name <- rownames(df)
df <- df[order(df$rsum,decreasing = TRUE),]
head(df, 10)
```

```{r,echo=FALSE}
write.table(df, paste0("../../results/top_transcripts/", project_ID, "_abundant_transcripts.txt"),
            quote = FALSE,
            row.names = FALSE)
```
