---
title: "Explore singlets"
author: "Kimberly Olney, PhD"
date: "September 2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
# Lewy body dementia center without walls LBD CWOW

# Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = "/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/")
setwd("/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/")
```

# Libraris, paths, colors
```{r echo=FALSE, message=FALSE}
source(here::here("/tgen_labs/jfryer/kolney/LBD_CWOW/LBD_snRNA/scripts/", "file_paths_and_colours.R"))
project_ID <- "CWOW_cellbender_singlets"
```

# Genes 
```{r gene_info}
if (file.exists("../../rObjects/annotation.rds")) {
  genes <- readRDS("../../rObjects/annotation.rds")
} else {
  gtf.file <- paste0(pathToRef, "/genes/genes.gtf")
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../rObjects/annotation.rds")
}

# gene table is an overview of the types of genes in the annotation file 
gene_type_table <- table(genes$gene_type)
write.table(gene_type_table, "gene_type_table.tsv", row.names = F, quote = F, sep = "\t")
write.table(genes, "genes.tsv", row.names = F, quote = F, sep = "\t")

# obtain the MT genes. Will be used for filtering down stream. 
mt.genes.df <- subset(genes, seqnames == "chrM")
genes_mt <- mt.genes.df$gene_name # 13 Mito genes 

rm(mt.genes.df, gene_type_table)
```

# QC thresholds
nCount_RNA = total number of transcripts (UMIs) in a single cell 
nFeature_RNA = number of unique genes (features)
```{r QC_metrics}
nCount.min <- 1000 
nCount.max <- 100000 
nFeature.min <- 800
nFeature.max<- 12000
cutoff_complexity <- 0.80
cutoff_mt <- 5
cutoff_hb <- 1
cutoff_ribo <- 1
cutoff_choroid <- 1
```

# Setup seurat object
```{r seurat_object}
dataObject <- readRDS(paste0("../../rObjects/", project_ID, ".rds"))
```

# Reorder samples
```{r reorder_samples}
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
dataObject$sample <- factor(sample, levels = order_samples)
table(dataObject$sample)  # check
Idents(dataObject) <- dataObject$sample
rm(sample, barcodes)
```

# Add metadata
```{r}
# Assuming 'metadata' is your metadata dataframe
# and its 'sampleID' column contains the cell IDs (the same IDs as the Seurat object's cells)
rownames(metadata) <- metadata$sampleID
# Reorder the metadata data frame to match the column order of the Seurat object
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
metadata_reordered <- metadata[sample, ]
# Add the entire metadata dataframe to the Seurat object
# Add the reordered metadata to the Seurat object
dataObject <- AddMetaData(object = dataObject, metadata = metadata_reordered)
dataObject$group <- factor(dataObject$TYPE, levels = c("CONTROL", "AD_AT", "LBD_S", "LBD_AS", "LBD_ATS"))
```

nuclei count per sample, sex, and group
```{r}
# Sample
table(dataObject$Sample_ID) 

# Sex 
table(dataObject$sex_inferred)

# Type
table(dataObject$group) 
```

QC columns 
```{r QC_columns}
summary(dataObject$nCount_RNA)
summary(dataObject$nFeature_RNA)
# cell.complexity
dataObject$cell.complexity <- log10(dataObject$nFeature_RNA) / log10(dataObject$nCount_RNA)

# Chromosome M
gene.names <- rownames(dataObject)
dataObject$percent.mt <- PercentageFeatureSet(dataObject, features = genes_mt)
summary(dataObject$percent.mt)

"^RPS|^RPL"
# ribosomal proteins 
ribo.genes <- gene.names[grep("^RPS|^RPL", gene.names)] 
mt.ribo <- gene.names[grep("^MRP[SL]", gene.names)]
ribo.combined <- c(mt.ribo,ribo.genes)
dataObject$percent.ribo <- PercentageFeatureSet(dataObject, features = ribo.combined)
summary(dataObject$percent.ribo)

# hemoglobin proteins
hb.genes <- gene.names[grep("^HB[BA]", gene.names)]
dataObject$percent.hb <- PercentageFeatureSet(dataObject, features = hb.genes)
summary(dataObject$percent.hb)

# percent choroid plexus
dataObject$percent.choroid <- PercentageFeatureSet(dataObject, features = c("TTR","FOLR1", "PRLR"))
summary(dataObject$percent.choroid)
```

# QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject$Sample_ID))
colnames(data) <- c("Sample_ID","frequency")

ncells1 <- ggplot(data, aes(x = Sample_ID, y = frequency, fill = Sample_ID)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            #vjust=-0.25, 
            hjust = -.025,
            angle = 90) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,30000, by = 2000), limits = c(0,30000)) +
  ggtitle("Nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1

rm(data)
```

```{r}
path <- paste0("../../results/nuclei_count/",project_ID, 
               "_cells_per_sample")
ncells1
saveToPDF(paste0(path, ".pdf"), width = 12, height = 6)
```

## Density plots
```{r prefiltering_density}
# set graphical parameter
par(mfrow = c(7,1))
# Visualize nCount_RNA
den1 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nCount_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) + theme(legend.position="none") +
  annotate("text", x = nCount.min, y = 0.1, label = paste("Min:", nCount.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nCount.max, y = 0.1, label = paste("Max:", nCount.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize nFeature
den2 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nFeature_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min) +
  geom_vline(xintercept = nFeature.max) + theme(legend.position="none") +
  annotate("text", x = nFeature.min, y = 0.1, label = paste("Min:", nFeature.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nFeature.max, y = 0.1, label = paste("Max:", nFeature.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = cell.complexity,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = cutoff_complexity) + theme(legend.position="none")  +
  annotate("text", x = cutoff_complexity, y = 0.1, label = paste("Min:", cutoff_complexity), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent.mt
den4 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.mt,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_mt) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_mt, y = 0.1, label = paste("Max:", cutoff_mt), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.ribo
den5 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.ribo,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_ribo) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Ribosomal Genes") +
  ylab("Density")+ theme(legend.position="none")  +
  annotate("text", x = cutoff_ribo, y = 0.1, label = paste("Max:", cutoff_ribo), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


# Visualize percent.hb
den6 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.hb,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_hb) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Hemoglobin Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_hb, y = 0.1, label = paste("Max:", cutoff_hb), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)


den7 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.choroid,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = cutoff_choroid) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Choroid Marker Genes") +
  ylab("Density")+ theme(legend.position="none") +
  annotate("text", x = cutoff_choroid, y = 0.1, label = paste("Max:", cutoff_choroid), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Arrange graphs in grid
plots1 <- list(den1,den2,den3,den4,den5,den6,den7)
layout1 <- rbind(c(1),c(2),c(3),c(4),c(5),c(6),c(7))
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

```{r}
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
path <- paste0("../../results/density/",project_ID,"_density")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 10)
dev.off()

rm(grid1, plots1)
```

## Violin plots
Warning: Default search for "data" layer in "RNA" assay yielded no results; utilizing "counts" layer instead."

This warning tells you that Seurat, by default, prefers to plot from the data (normalized) slot. However, we haven't normalized the RNA assay, thus the data slot is empty. As a fallback, Seurat  switched to using the counts slot, which contains the raw UMI counts. 
```{r prefiltering_violins}
# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(dataObject,
              features = c( "nCount_RNA","nFeature_RNA", "cell.complexity"),
              ncol = 1,
              group.by = 'Sample_ID',
              ##cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(dataObject,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 2,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v2
```


```{r, echo=FALSE}
# save v1
v1
path <- paste0("../../results/violin/",project_ID,"_nFeature_nCount_complexity")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 9)
dev.off()

# save v2
v2
path <- paste0("../../results/violin/",project_ID,"_percent")
saveToPDF(paste0(path, ".pdf"), width = 16, height = 7)
dev.off()

# cleanup
#remove(v1,v2)
```

